nalgebra-sparse/src/cs.rs

```diff
@@ -630,56 +631,63 @@ where
                         return Err(SparseFormatError::from_kind_and_msg(
                             SparseFormatErrorKind::DuplicateEntry,
                             "Input data contains duplicate entries.",
                         ));
                     }
-                    if prev > next {
+                    if prev > minor_idx {
+                        if !sort {
+                            return Err(SparseFormatError::from_kind_and_msg(
+                                SparseFormatErrorKind::InvalidStructure,
+                                "Minor indices are not monotonically increasing within each lane.",
+                            ));
+                        }
                         nonmonotonic = true;
                     }
                 }
-                prev = Some(next);
+                prev = Some(minor_idx);
             }
 
             // sort if indices are nonmonotonic and sorting is expected
             if nonmonotonic && sort {
-                minor_index_permutation.resize(range_end - range_start, 0);
-                for (index, value) in (range_start..range_end).enumerate() {
-                    minor_index_permutation[index] = value;
+                let range_size = range_end - range_start;
+                minor_index_permutation.resize(range_size, 0);
+                compute_sort_permutation(&mut minor_index_permutation, &minor_idx_in_lane);
+                minor_idx_buffer.resize(range_size, 0);
+                for (index, &value) in minor_idx_in_lane.iter().enumerate() {
+                    minor_idx_buffer[index] = value;
                 }
-
-                // sort permutation values for monotonic index order
-                minor_index_permutation.sort_by(|a, b| {
-                    let x = &minor_idx_in_lane[*a - range_start];
-                    let y = &minor_idx_in_lane[*b - range_start];
-                    x.partial_cmp(y).unwrap()
-                });
-                let unsorted_indices: Vec<usize> = minor_indices[range_start..range_end]
-                    .iter()
-                    .map(|i| i.clone())
-                    .collect();
-                let unsorted_values: Vec<T> =
-                    Vec::from_iter(((range_start..range_end).map(|i| &value_refs[i])).cloned());
-                for (index, &offset) in minor_index_permutation.iter().enumerate() {
-                    let i: usize = unsorted_indices[offset - range_start];
-                    minor_indices[index + range_start] = i;
-                    let v: T = unsorted_values[offset - range_start].clone();
-                    value_refs[index + range_start] = v;
+                apply_permutation(
+                    &mut minor_indices[range_start..range_end],
+                    &minor_idx_buffer,
+                    &minor_index_permutation,
+                );
+
+                // sort values if they exist
+                if !value_refs.is_empty() {
```

Comment:
This is pretty confusing, as it conflates whether or not the set of values is empty with whether or not they are present. Perhaps use `if let`?

```rust
if let Some(values) = values {
    let mut values_in_lane = ...;
    values_buffer.clear();
    values_buffer.extend_from_slice(&values_in_lane);
    apply_permutation(...)
}
```
I think this way we can also lift the `T: Zero` requirement since we no longer need to fill the buffer with arbitrary values.

Reply:
I think I already tried this but there is an issue as the sort is applied only to a range of values and can be run in every iteration of the loop. So I get following error message if I try it: "_value moved here, in previous iteration of loop_".
Nevertheless, we can still lift T: Zero, as mentioned in my comment above.

Maybe we can find a better approach of checking values here.
But then another question comes up. How should handling of absent values differ from handling an empty list of values?

Reply:
Ah, right, this is a bit tricky. The reason is that you would consume the `Option` at every iteration. I think this should work:
```rust
if let Some(values) = values.as_mut() {
   
}
```

The difference is mainly in semantics: The logic may *technically* be identical (not 100% sure), but when reading the code it's not quite clear what's going on. Basically, "values not provided" and "values provided but empty" are two different cases, and it can get a bit confusing if we batch them together.

Reply:
Indeed, this works if we use a mutable variable for the values option check!